File: skills/code-architect/references/code-examples/typescript/Singleton.ts
// Singleton — ensure one instance with a global access point.

class ConfigRegistry {
  private static instance: ConfigRegistry | null = null;
  private values = new Map<string, string>();

  // Private constructor blocks external instantiation.
  private constructor() {}

  static getInstance(): ConfigRegistry {
    if (ConfigRegistry.instance === null) {
      ConfigRegistry.instance = new ConfigRegistry();
    }
    return ConfigRegistry.instance;
  }

  set(key: string, value: string): void {
    this.values.set(key, value);
  }

  get(key: string): string | undefined {
    return this.values.get(key);
  }
}

function demo(): void {
  ConfigRegistry.getInstance().set("region", "eu-west");
  const sameInstance = ConfigRegistry.getInstance();
  console.log(`region = ${sameInstance.get("region")}`);
  console.log(`same instance: ${ConfigRegistry.getInstance() === sameInstance}`);
}

demo();

File: skills/code-architect/references/code-examples/python/singleton.py
"""Singleton — ensure one instance with a global access point."""

from __future__ import annotations

import threading


class Registry:
    """Thread-safe single instance reached through get_instance()."""

    _instance: "Registry | None" = None
    _lock = threading.Lock()

    def __init__(self) -> None:
        self._entries: dict[str, str] = {}

    @classmethod
    def get_instance(cls) -> "Registry":
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = cls()
        return cls._instance

    def set(self, key: str, value: str) -> None:
        self._entries[key] = value

    def get(self, key: str) -> str | None:
        return self._entries.get(key)


if __name__ == "__main__":
    a = Registry.get_instance()
    a.set("region", "eu-west")
    b = Registry.get_instance()
    print("same object:", a is b)
    print("shared value:", b.get("region"))
